autopush/base.py

```diff
@@ -1,8 +1,122 @@
+PREF_SCHEME = "webpush"
+
+DEFAULT_ERR_URL = ("http://autopush.readthedocs.io/en/latest/http.html"
+                   "#error-codes")
+
+
+class ThreadedValidate(object):
+    """A cyclone request validation decorator
+
+    Exposed as a classmethod for running a marshmallow-based validation schema
+    in a separate thread for a cyclone request handler.
+
+    """
+    log = Logger()
+
+    def __init__(self, schema):
+        self.schema = schema
+
+    def _validate_request(self, request_handler):
+        """Validates a schema_class against a cyclone request"""
+        data = {
+            "headers": request_handler.request.headers,
+            "body": request_handler.request.body,
+            "path_args": request_handler.path_args,
+            "path_kwargs": request_handler.path_kwargs,
+            "arguments": request_handler.request.arguments,
+        }
+        schema = self.schema()
+        schema.context["settings"] = request_handler.ap_settings
+        schema.context["log"] = self.log
+        return schema.load(data)
+
+    def _call_func(self, result, func, request_handler, *args, **kwargs):
+        output, errors = result
+        if errors:
+            request_handler._write_validation_err(errors)
+        else:
+            request_handler.valid_input = output
+            return func(request_handler, *args, **kwargs)
+
+    def _decorator(self, func):
+        @wraps(func)
+        def wrapper(request_handler, *args, **kwargs):
+            # Wrap the handler in @cyclone.web.synchronous
+            request_handler._auto_finish = False
+            d = deferToThread(self._validate_request, request_handler)
+            d.addCallback(self._call_func, func, request_handler, *args,
+                          **kwargs)
+            d.addErrback(request_handler._overload_err)
+            d.addErrback(request_handler._boto_err)
+            d.addErrback(request_handler._validation_err)
+            d.addErrback(request_handler._response_err)
+        return wrapper
+
+    @classmethod
+    def validate(cls, schema):
+        """Validate a request schema in a separate thread before calling the
+        request handler
+
+        An alias `threaded_validate` should be used from this module.
+
+        Using `cyclone.web.asynchronous` is not needed as this function
+        will attach equivilant functionality to the method handler. Calling
+        `self.finish()` is needed on decorated handlers.
+
+        .. code-block::
+
+            class MyHandler(cyclone.web.RequestHandler):
+                @threaded_validate(MySchema())
+                def post(self):
+                    ...
+
+        """
+        return cls(schema)._decorator
+
+
+# Alias to the validation classmethod decorator
+threaded_validate = ThreadedValidate.validate
+
+
+@attrs
+class Notification(object):
+    """Parsed notification from the request"""
+    version = attrib()
+    data = attrib()
+    channel_id = attrib()
```

Comment:
If we're moving this around, maybe it should join WebPushNotification in utils? Or its own module?